### ISSUE TITLE
Did you convert the range map to a Z Map?

### ISSUE THREAD

**Issue by soulslicer** [uos/basler_tof#2]

**Comment by mintar** [comment]
Text?

**Comment by soulslicer** [comment]
Did you convert the range map to a Z Map?

**Comment by mintar** [comment]
Could you please elaborate?

**Comment by soulslicer** [comment]
Well..ROS programs consume depth maps, but from what i see, you output the depth map directly from the Basler API which is a Range Map

**Comment by mintar** [comment]
As far as I remember, the Basler API also returns a depth image, so we can immediately convert that into a depth image. It's almost a year ago since I last tested this, but I remember that I calibrated the depth image using [camera_calibration](http://wiki.ros.org/camera_calibration) and then converted it to a point cloud using [depth_image_proc](http://wiki.ros.org/depth_image_proc), and the resulting point cloud was correct. This means that the driver outputs the correct depth image (except if Basler recently changed their SDK...).

Actually, the launch file included with this driver does exactly what I have just described, so you can check for yourself.

**Comment by soulslicer** [comment]
No but in your code it says you have selected range. And your node is outputting a range map. If you pass it into depth_image_proc, the point cloud will be wrong

**Comment by mintar** [comment]
I've just tried it, and this is what I get:

![basler](https://user-images.githubusercontent.com/320188/33762694-bf01aae4-dc0d-11e7-9a10-e2111a0859ce.png)

The yellow point cloud is the one computed from the depth image, the other one is the one published by the driver directly. The yellow one is slightly curved (especially the wall on the left isn't straight), but it has to be due to the calibration I did being incorrect. **Crucially, the Z values of the green point cloud and the depth image are the same.**

### Basler documentation for ComponentSelector "Range":

A depth image is acquired. Each pixel in the image contains the information of the distance between the camera and the object. The way the distance is represented depends on the option specified with the Pixel Format parameter, either a range image or a point cloud.


### Basler documentation for PixelFormat:

**Coord3D_C16:** The pixel format used is 3D coordinate C 16-bit. The result is a range image where each pixel contains distance data.

**Coord3D_ABC32f:** The pixel format used is 3D coordinate ABC 32-bit floating point. The result is a point cloud where the origin of each point in the cloud can be traced back to the pixel that generated the point.

----------------------------------------------------------------------

This is what I do in the driver:

```c++
ptrImageComponentSelector->FromString("Range");
ptrImageComponentEnable->SetValue(true);
ptrPixelFormat->FromString("Coord3D_ABC32f");

// [...]

depth_cvimg.image.at<uint16_t>(i, j) = static_cast<uint16_t>(pPoint->z);
```

In other words, the depth image that my driver publishes contains Z values, not range values. This is why I believe it's correct.

**Comment by soulslicer** [comment]
Ok..

Because I tried it and it didn't work. Take the depth map your driver generates, and pass it into the point cloud generator in depth_image_proc modules. Point the camera straight at a flat wall at least 5m away. Tell me if it's actually straight.

**Comment by mintar** [comment]
> Because I tried it and it didn't work. Take the depth map your driver generates, and pass it into the point cloud generator in depth_image_proc modules. Point the camera straight at a flat wall at least 5m away. Tell me if it's actually flat.

No, it isn't, as you can see from the picture I posted. The wall in the green point cloud (= the one from the driver) is straight, the same wall in the yellow point cloud (= the one computed from the depth image) is not. I agree with you that the green point cloud is correct and the yellow one isn't. We only disagree on the explanation.

> It won't be and it will be curved. Because it's a Range map. And it's not because of your calibration

What makes you think it's not because of my calibration?

> Range Map != Depth Map

Exactly. Range = distance from camera, depth = perpendicular distance to the virtual camera plane = z coordinate of the point cloud. As I explained above, the depth image returned by the driver has the z coordinate. This is the same value (`pPoint->z`) that I put into the green point cloud, so it **cannot be a range**.

I'm not 100% sure *why* the point cloud calculated from my depth image is curved. It might be the calibration, as I've speculated above, or it might be some other bug in my code (probably in the camera info that goes into depth_image_proc). All I'm saying is that the depth image I'm publishing is correct.

If you find the bug, please let me know. If you still believe that the depth image should be converted somehow, try implementing it and see where it gets you.
